### PR TITLE
Track token list updates

### PR DESCRIPTION
**Pull request by Velenir**
Adds subscription to TokenListApi so that updates propagate to everywhere we need tokenList

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr779--dexreact.review.gnosisdev.com

File: src/components/TradeWidget/index.tsx [hunk 1]
```diff
@@ -356,6 +356,8 @@ const TradeWidget: React.FC = () => {
 
   // If user is connected, use balances, otherwise get the default list
   const tokens = useMemo(
+    // it's okay to tokenListApi.getTokens() here without subscribing to updates
+    // because balances from useTokenBalances is already subscribed
     () => (isConnected && balances.length > 0 ? balances : tokenListApi.getTokens(fallBackNetworkId)),
```
Comment (alfetopito) [inline src/components/TradeWidget/index.tsx]: But, balances is only used/updated if wallet is connected.
Won't we allow tokens to be added on trade page without having connected the wallet?

I guess that is fine for adding tokens, but can be a problem due to the way I'll load the tokens. (can probably address on my PR, though)

File: src/hooks/useTokeList.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// for stable reference
+// to avoid updates on setState([])
+const emptyArray: TokenDetails[] = []
+
+export const useTokenList = (networkId?: number): TokenDetails[] => {
+  const [tokens, setTokens] = useSafeState(networkId === undefined ? emptyArray : tokenListApi.getTokens(networkId))
+
+  useEffect(() => {
+    if (networkId === undefined) return setTokens(emptyArray)
+    return tokenListApi.subscribe(setTokens)
```
Comment (alfetopito) [inline src/hooks/useTokeList.ts]: :thinking: 
When the hook umounts, shouldn't it return `tokenListApi.unsubscribe`?

File: src/services/index.ts [hunk 3]
```diff
@@ -7,7 +7,7 @@ import {
   getPriceEstimationFactory,
 } from './factories'
 import { logDebug } from 'utils'
-import { TokenDetails } from '@gnosis.pm/dex-js'
+import { TokenDetails } from 'types'
```
Comment (alfetopito) [inline src/services/index.ts]: Are they different?

File: src/api/tokenList/Subscriptions.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+type SubscriptionCallback<T> = (newState: T) => void
+
+export interface SubscriptionsInterface<State> {
+  subscribe(callback: SubscriptionCallback<State>): Command
+  unsubscribe(callback: SubscriptionCallback<State>): void
+  unsubscribeAll(): void
+  triggerSubscriptions(newState: State): void
+}
+
+export default class GenericSubscriptions<State> implements SubscriptionsInterface<State> {
```
Comment (anxolin) [inline src/api/tokenList/Subscriptions.ts]: One idea. Why don't you create a new interface `Subscriptions<State>` and then do `GenericSubscriptions` implement this interface plus add the public `triggerSubscriptions` method.

This way TokenList would extend Subscriptions, and the Impl/Mock would extend GenericSubscriptions

File: src/hooks/useTokenBalances.ts [hunk 5]
```diff
@@ -12,6 +12,7 @@ import { ALLOWANCE_FOR_ENABLED_TOKEN } from 'const'
 import { TokenBalanceDetails, TokenDetails } from 'types'
 import { WalletInfo } from 'api/wallet/WalletApi'
 import { PendingFlux } from 'api/deposit/DepositApi'
+import { useTokenList } from './useTokeList'
```
Comment (W3stside) [inline src/hooks/useTokenBalances.ts]: `useTokeList` typo error

Comment (W3stside) [reply src/hooks/useTokenBalances.ts]: unless it's a list of you smoking doobies idk

File: src/components/TradeWidget/index.tsx (hunk 1 above)
Comment (Velenir) [reply src/components/TradeWidget/index.tsx]: ugh, you are right
having default network-less tokenList is a problem

File: src/hooks/useTokeList.ts (hunk 2 above)
Comment (Velenir) [reply src/hooks/useTokeList.ts]:
```ts
public subscribe(callback: SubscriptionCallback<State>): Command {
    this._subscriptions.push(callback)

    return (): void => this.unsubscribe(callback)
  }
```

File: src/services/index.ts (hunk 3 above)
Comment (Velenir) [reply src/services/index.ts]:
```ts
//dex-js
export interface TokenDetails {
    id: number;
    name?: string;
    symbol?: string;
    decimals?: number;
    address: string;
    addressMainnet?: string;
    image?: string;
}

//dex-react
export interface MinimalTokenDetails {
  address: string
  symbol?: string
  name?: string
  decimals: number
}

export interface TokenDetails extends MinimalTokenDetails {
  id: number
  addressMainnet?: string
  image?: string
}
```

`decimals` is required in dex-react
which probably doesn't matter much, but TS was complaining

File: src/api/tokenList/Subscriptions.ts (hunk 4 above)
Comment (Velenir) [reply src/api/tokenList/Subscriptions.ts]: Can't extend interface
Can only extend class `GenericSubscriptions`, so `triggerSubscriptions` becomes public, so we remove it from interface with 
```ts
export interface TokenList extends Omit<SubscriptionsInterface<TokenDetails[]>, 'triggerSubscriptions'>
```

But you gave me an idea, I'll make it `protected`, 6abef37b3728105c8fe616875be650b913c81729

**Comment by Velenir** [comment]
I'll merge this and #760 to unblock @alfetopito

File: src/hooks/useTokeList.ts (hunk 2 above)
Comment (alfetopito) [reply src/hooks/useTokeList.ts]: :brainfreeze:

**Pull request merged**